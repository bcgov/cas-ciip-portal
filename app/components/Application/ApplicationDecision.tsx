import React from 'react';
import {Alert} from 'react-bootstrap';
import {getUserFriendlyStatusLabel} from 'lib/text-transforms';

interface Props {
  actionRequired: Boolean;
  decision: string;
  reviewComments: string[];
}

const BS_ALERT = {
  REQUESTED_CHANGES: 'warning',
  REJECTED: 'danger',
  APPROVED: 'success'
};

const DECISION_TEXT = {
  REQUESTED_CHANGES: '',
  REJECTED:
    'Thank you for applying to the CIIP. Unfortunately this submission did not meet program requirements and a payment cannot be granted.',
  APPROVED:
    'This application has been approved. Your incentive payment is being processed. No further action is required from you at this time.'
};

const ApplicationDecision: React.FunctionComponent<Props> = ({
  actionRequired,
  decision,
  reviewComments,
  children
}) => {
  const heading = actionRequired
    ? 'Action Required'
    : decision === 'REQUESTED_CHANGES'
    ? 'Changes Requested'
    : `Decision: ${getUserFriendlyStatusLabel(decision)}`;

  // Items will not be re-ordered after mounting, so using index is safe
  // eslint-disable-next-line react/no-array-index-key
  const items = reviewComments.map((c, i) => <li key={`comment-${i}`}>{c}</li>);

  return (
    <Alert variant={actionRequired ? 'danger' : BS_ALERT[decision]}>
      <Alert.Heading>{heading}</Alert.Heading>
      <p>
        {actionRequired
          ? 'Changes to your application have been requested. Please read these review notes and gather the relevant information before revising your application.'
          : DECISION_TEXT[decision]}
      </p>
      {reviewComments.length > 0 && (
        <>
          <hr />
          <p className="h5">Reviewer Notes:</p>
          <ul>{items}</ul>
        </>
      )}
      {children}
    </Alert>
  );
};

export default ApplicationDecision;
