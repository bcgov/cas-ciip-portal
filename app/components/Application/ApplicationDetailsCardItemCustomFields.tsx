import React from 'react';
import SummaryEmissionGasFields from 'containers/Forms/SummaryEmissionGasFields';
import SummaryEmissionSourceFields from 'containers/Forms/SummaryEmissionSourceFields';
import ProductionFields from 'containers/Forms/ProductionFields';
import {FieldProps} from 'react-jsonschema-form';
import NumberFormat from 'react-number-format';

const customFields = (
  showDiff,
  diffPathArray,
  diffArray,
  handleEnums,
  previousIsEmpty
) => {
  const CUSTOM_FIELDS: Record<string, React.FunctionComponent<FieldProps>> = {
    TitleField: props => <h3>{props.title}</h3>,
    StringField: props => {
      let prevValue;
      let hasDiff = false;
      if (showDiff) {
        hasDiff = diffPathArray.includes(
          props.idSchema.$id.replace(/^root_/g, '')
        );
        prevValue =
          diffArray[
            diffPathArray.indexOf(props.idSchema.$id.replace(/^root_/g, ''))
          ];

        if ((hasDiff || previousIsEmpty) && prevValue !== props.formData) {
          prevValue = handleEnums(props, false, prevValue);
          const currentValue = handleEnums(props, true, prevValue);

          return (
            <>
              <span style={{backgroundColor: '#ffeef0'}}>
                {prevValue ? prevValue : <i>[No Data Entered]</i>}
              </span>
              &nbsp;---&gt;&nbsp;
              <span style={{backgroundColor: '#e6ffed'}}>
                {currentValue ? currentValue : <i>[No Data Entered]</i>}
              </span>
            </>
          );
        }
      }

      if (
        props.formData === null ||
        props.formData === undefined ||
        props.formData === ''
      )
        return <i>[No Data Entered]</i>;

      const value = handleEnums(props, true, prevValue);
      return value;
    },
    BooleanField: props => {
      const hasDiff = diffPathArray.includes(
        props.idSchema.$id.replace(/^root_/g, '')
      );

      if (showDiff && hasDiff) {
        const prevValue =
          diffArray[
            diffPathArray.indexOf(props.idSchema.$id.replace(/^root_/g, ''))
          ];
        return (
          <>
            <span style={{backgroundColor: '#ffeef0'}}>
              {prevValue ? 'Yes' : 'No'}
            </span>
            &nbsp;---&gt;&nbsp;
            <span style={{backgroundColor: '#e6ffed'}}>
              {props.formData ? 'Yes' : 'No'}
            </span>
          </>
        );
      }

      return <>{props.formData ? 'Yes' : 'No'} </>;
    },
    emissionSource: props => <SummaryEmissionSourceFields {...props} />,
    emissionGas: props => <SummaryEmissionGasFields {...props} />,
    production: props => (
      <ProductionFields query={props.formContext.query} {...props} />
    ),
    NumberField: props => {
      if (
        props.formData === null ||
        props.formData === undefined ||
        props.formData === ''
      )
        return <i>[No Data Entered]</i>;

      let prevValue;
      let hasDiff = false;
      if (showDiff) {
        hasDiff = diffPathArray.includes(
          props.idSchema.$id.replace(/^root_/g, '')
        );
        prevValue =
          diffArray[
            diffPathArray.indexOf(props.idSchema.$id.replace(/^root_/g, ''))
          ];
        if (hasDiff || previousIsEmpty) {
          prevValue = handleEnums(props, false, prevValue);
          const currentValue = handleEnums(props, true, prevValue);

          return (
            <>
              <span style={{backgroundColor: '#ffeef0'}}>
                {prevValue ? (
                  <NumberFormat
                    thousandSeparator
                    displayType="text"
                    value={prevValue}
                  />
                ) : (
                  <i>[No Data Entered]</i>
                )}
              </span>
              &nbsp;---&gt;&nbsp;
              <span style={{backgroundColor: '#e6ffed'}}>
                {currentValue ? (
                  <NumberFormat
                    thousandSeparator
                    displayType="text"
                    value={currentValue}
                  />
                ) : (
                  <i>[No Data Entered]</i>
                )}
              </span>
            </>
          );
        }
      }

      const value = handleEnums(props, true, prevValue);
      return (
        <NumberFormat thousandSeparator displayType="text" value={value} />
      );
    }
  };
  return CUSTOM_FIELDS;
};

export default customFields;
