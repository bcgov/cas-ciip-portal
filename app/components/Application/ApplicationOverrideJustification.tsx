import React, {useState, useEffect} from 'react';
import {createFragmentContainer, RelayProp} from 'react-relay';
import {Accordion, Alert, Button, Form} from 'react-bootstrap';
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import {faCaretDown} from '@fortawesome/free-solid-svg-icons';
import updateApplicationRevisionMutation from 'mutations/application/updateApplicationRevisionMutation';

interface Props {
  overrideActive: boolean;
  setOverrideActive: (boolean) => void;
  applicationOverrideJustification: string;
  revisionId: string;
  relay: RelayProp;
  hasErrors: any;
}

export const ApplicationOverrideJustificationComponent: React.FunctionComponent<Props> = ({
  overrideActive,
  setOverrideActive,
  applicationOverrideJustification,
  revisionId,
  relay,
  hasErrors
}) => {
  const [accordionOpen, setAccordionOpen] = useState(false);
  const [overrideJustification, setOverrideJustification] = useState(
    applicationOverrideJustification
  );
  const [emptyError, setEmptyError] = useState('');

  const handleOverrideChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    setOverrideJustification(e.target.value);
  };

  const handleOverrideCancel = () => {
    setAccordionOpen(false);
    setEmptyError('');
    setOverrideActive(applicationOverrideJustification !== null);
    setOverrideJustification(applicationOverrideJustification || null);
  };

  const handleOverrideDelete = async () => {
    const {environment} = relay;
    const variables = {
      input: {
        id: revisionId,
        applicationRevisionPatch: {
          overrideJustification: null
        }
      }
    };

    await updateApplicationRevisionMutation(environment, variables);
    setOverrideJustification(null);
    setOverrideActive(false);
  };

  const handleOverrideSave = async () => {
    if (overrideJustification) {
      setAccordionOpen(false);
      const {environment} = relay;
      const variables = {
        input: {
          id: revisionId,
          applicationRevisionPatch: {
            overrideJustification
          }
        }
      };

      await updateApplicationRevisionMutation(environment, variables);
      setEmptyError('');
      setOverrideActive(true);
    } else setEmptyError('Justification cannot be empty');
  };

  const handleOverrideEdit = () => {
    setOverrideActive(false);
    setAccordionOpen(true);
  };

  useEffect(() => {
    if (overrideJustification && hasErrors === false) handleOverrideDelete();
  });

  const errorsWithOverrideActive = (
    <>
      <Alert variant="danger">
        <strong>Error Validation Override Active</strong>
      </Alert>
      <Alert variant="secondary">
        <p>
          <strong>
            You have chosen to override the errors present in your application.
            This may cause a delay in processing your application.
          </strong>
        </p>
        <p>
          <strong>Your Override Justification:</strong>
        </p>
        <p>{overrideJustification}</p>
        <Button
          style={{marginRight: '5px'}}
          variant="secondary"
          onClick={handleOverrideEdit}
        >
          Edit Justification
        </Button>
        <Button variant="danger" onClick={handleOverrideDelete}>
          Delete Override
        </Button>
      </Alert>
    </>
  );

  const errorsWithoutOverrideActive = (
    <>
      <Alert variant="danger">
        <Accordion>
          <div className="override-accordion">
            Your application contains errors shown below that must be fixed
            before submission. You may either correct these or alternatively,
            override and provide justification. Your justification will be
            reviewed by CAS and may cause a delay in processing your
            application.
            <Accordion.Toggle
              as={Button}
              variant="secondary"
              eventKey="0"
              onClick={() => setAccordionOpen(!accordionOpen)}
            >
              Override and Justify
              <FontAwesomeIcon
                icon={faCaretDown}
                style={{marginLeft: '0.75em'}}
              />
            </Accordion.Toggle>
          </div>
          <Accordion.Collapse in={accordionOpen} eventKey="0">
            <>
              <Form>
                <h4>Override Form Validation</h4>
                <Form.Group controlId="overrideJustification">
                  <Form.Label>Justification for incomplete form:</Form.Label>
                  <Form.Control
                    as="textarea"
                    rows={4}
                    value={overrideJustification || ''}
                    onChange={handleOverrideChange}
                  />
                </Form.Group>
                <Button variant="success" onClick={handleOverrideSave}>
                  Save
                </Button>
                <Accordion.Toggle
                  as={Button}
                  eventKey="0"
                  variant="light"
                  style={{
                    marginLeft: '1em',
                    border: '1px solid currentColor'
                  }}
                  onClick={handleOverrideCancel}
                >
                  Cancel
                </Accordion.Toggle>
              </Form>
              {!overrideJustification && (
                <p className="justification-text">{emptyError}</p>
              )}
            </>
          </Accordion.Collapse>
        </Accordion>
      </Alert>
      <style jsx>
        {`
          .override-accordion {
            display: flex;
            flex-wrap: wrap;
            justify-content: flex-end;
          }
          .justification-text {
            color: red;
          }
        `}
      </style>
    </>
  );

  const renderOverride = () => {
    if (hasErrors === false) return null;
    if (hasErrors && overrideActive) return errorsWithOverrideActive;
    return errorsWithoutOverrideActive;
  };

  return renderOverride();
};

export default createFragmentContainer(
  ApplicationOverrideJustificationComponent,
  {}
);
