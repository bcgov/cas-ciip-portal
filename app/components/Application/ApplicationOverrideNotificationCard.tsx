import React from 'react';
import {Card} from 'react-bootstrap';

interface Props {
  overrideJustification: string;
}

const ApplicationOverrideNotification: React.FunctionComponent<Props> = ({
  overrideJustification
}) => {
  if (overrideJustification)
    return (
      <Card id="justification-card">
        <Card.Header className="bg-danger">
          Error validation override is active
        </Card.Header>
        <Card.Body tabIndex={0}>
          <Card.Text>
            <strong>
              There were errors in this application that the reporter has chosen
              to override. Justification for overriding the errors is provided
              in the box below:
            </strong>
          </Card.Text>
          <Card>
            <div id="justification" className="scrollbar">
              <Card.Body>{overrideJustification}</Card.Body>
            </div>
          </Card>
        </Card.Body>
        <style jsx>{`
          #justification-card {
            margin: 1rem 0;
          }
          #justification {
            max-height: 19.2em;
            background: #eee;
            overflow-y: scroll;
          }
          .scrollbar::-webkit-scrollbar {
            -webkit-appearance: none;
            width: 8px;
          }
          .scrollbar::-webkit-scrollbar-thumb {
            border-radius: 4px;
            background-color: rgba(0, 0, 0, 0.5);
            box-shadow: 0 0 1px rgba(255, 255, 255, 0.5);
          }
        `}</style>
      </Card>
    );
};

export default ApplicationOverrideNotification;
