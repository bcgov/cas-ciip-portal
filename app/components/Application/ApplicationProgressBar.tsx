import React from "react";
import ProgressStepIndicator from "components/ProgressStepIndicator";
import StatusBadgeColor from "components/helpers/StatusBadgeColor";

interface Props {
  title: string;
  ariaLabel?: string; // used if the indicator does not have a title.
  completed?: number;
}

const ciipAppSteps = [
  {
    description: "CIIP Application not started",
    badgeStyle: StatusBadgeColor.NONE,
    number: 1,
  },
  {
    description: "Application in progress",
    badgeStyle: StatusBadgeColor.INITIAL,
    number: 2,
  },
  {
    description: "Application submitted and in review",
    badgeStyle: StatusBadgeColor.PENDING,
    number: 3,
  },
  {
    description: "Application approved or rejected",
    badgeStyle: StatusBadgeColor.APPROVED,
    number: 4,
  },
];

export const ApplicationProgressBar: React.FunctionComponent<Props> = ({
  completed = 0,
  ...props
}) => {
  const completedSteps = ciipAppSteps.map((step, i) => {
    return completed > i ? { ...step, completed: true } : step;
  });
  return <ProgressStepIndicator steps={completedSteps} {...props} />;
};

export default ApplicationProgressBar;
