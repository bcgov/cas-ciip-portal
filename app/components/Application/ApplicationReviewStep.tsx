import React from 'react';
import {Button, Card} from 'react-bootstrap';
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import {faCheck} from '@fortawesome/free-solid-svg-icons';

interface Props {
  reviewStep: string;
}

export const ApplicationReviewStep: React.FunctionComponent<Props> = ({
  reviewStep
}) => {
  return (
    <>
      <Card className="text-center justification-card">
        <div id="justification" className="scrollbar">
          <h1>{reviewStep} Review</h1>
          <Button style={{width: '80%'}} variant="outline-info">
            <FontAwesomeIcon icon={faCheck} />
            Mark this step as complete
          </Button>
          <Card.Body className="scrollbar">
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE COMMENTS GO HERE
            COMMENTS GO HERE
          </Card.Body>
        </div>
        <Card.Footer className="text-muted">2 days ago</Card.Footer>
      </Card>
      <style jsx>{`
        #justification-card {
          margin: 1rem 1rem 1rem 1rem;
        }
        #justification {
          position: fixed;
          padding: 1rem;
          max-height: 50em;
          background: #eee;
          overflow-y: scroll;
          align: center;
        }
        .scrollbar::-webkit-scrollbar {
          -webkit-appearance: none;
          width: 8px;
        }
        .scrollbar::-webkit-scrollbar-thumb {
          border-radius: 4px;
          background-color: rgba(0, 0, 0, 0.5);
          box-shadow: 0 0 1px rgba(255, 255, 255, 0.5);
        }
      `}</style>
    </>
  );
};

export default ApplicationReviewStep;
