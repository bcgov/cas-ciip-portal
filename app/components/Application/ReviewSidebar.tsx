import React from 'react';
import {Button} from 'react-bootstrap';
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import {faCheck} from '@fortawesome/free-solid-svg-icons';

interface Props {
  reviewStep: string;
  onClose: () => void;
  onCompleted: () => void;
}

export const ApplicationReviewStep: React.FunctionComponent<Props> = ({
  reviewStep,
  onClose,
  onCompleted
}) => {
  return (
    <div id="sidebar" className="col-md-4 col-xxl-3">
      <button
        type="button"
        id="close"
        aria-label={`Close ${reviewStep} Review`}
        onClick={onClose}
      >
        Ã—
      </button>
      <h2>{reviewStep} Review</h2>
      <Button
        variant="outline-primary"
        type="button"
        onClick={onCompleted}
        style={{
          padding: '0.75rem .9rem',
          display: 'block',
          margin: 'auto'
        }}
      >
        <FontAwesomeIcon icon={faCheck} style={{marginRight: '0.5rem'}} />
        Mark this review step completed
      </Button>
      <div id="scrollable-comments" tabIndex={0}>
        HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO
        HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO
        HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO
        HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO
        HELLO HELLO sd sdjf sdjfh lk ie dsdfo o od pd j h y ii k k i i i y y y y
        j j n n b g g g g g j u i k j n k l m n b b v v h k i k j n k l m n b b
        v v h k h k i k j n k l m n b b v v h k i k j n k l m n b b v v h k i k
        j n k l m n b b v v h k i k j n k l m n b b v v h k i k j n k l m n b b
        v v h k i k j n k l m n b b v v h k i k j n k l m n b b v v h k i k j n
        k l m n b b v v h k i k j n k l m n b b v v h k i k j n k l m n b b v v
        h k i k j n k l m n b b v v h k i k j n k l m n b b v v h k i k j n k l
        m n b b v v h k i k j n k l m n b b v v h k i k j n k l m n b b v v h k
        i k j n k l m n b b v v h k i k j n k l m n b b v v h k i k j n k l m n
        b b v v h k i k j n k l m n b b v v h k i k j n k l m n b b v v h k i k
        j n k l m n b b v v
      </div>
      <div id="button-footer">
        <Button variant="link" style={{padding: 0}}>
          Show Resolved
        </Button>
        <Button variant="primary">+ New Comment</Button>
      </div>
      <style jsx>{`
        #sidebar {
          position: fixed;
          top: 68px;
          right: 0;
          height: calc(100% - 68px);
          background: #f7f7f7;
          border: 1px solid transparent;
          border-left-color: #dfdfdf;
          padding: 0 1.5em;
        }
        #close {
          font-size: 2.2rem;
          position: absolute;
          top: 0;
          left: 1rem;
          cursor: pointer;
          border: none;
          background: none;
        }
        h2 {
          margin: 15px 0;
          text-align: center;
        }
        #scrollable-comments {
          overflow-y: scroll;
          height: calc(100% - 108px - 3rem - 53px);
          margin: 1.5rem 0;
          padding: 0.5em 0.7em 0 0.5em;
        }
        #scrollable-comments::-webkit-scrollbar {
          -webkit-appearance: none;
          width: 8px;
        }
        #scrollable-comments::-webkit-scrollbar-thumb {
          border-radius: 4px;
          background-color: rgba(0, 0, 0, 0.5);
          box-shadow: 0 0 1px rgba(255, 255, 255, 0.5);
        }
        #button-footer {
          position: absolute;
          bottom: 0;
          margin-bottom: 15px;
          width: calc(100% - 3em);
          display: flex;
          justify-content: space-between;
        }
      `}</style>
    </div>
  );
};

export default ApplicationReviewStep;
