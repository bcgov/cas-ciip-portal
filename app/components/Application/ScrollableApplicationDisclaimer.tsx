import React from "react";
import { Card } from "react-bootstrap";
import LegalDisclaimerText from "components/LegalDisclaimerText";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faExternalLinkAlt } from "@fortawesome/free-solid-svg-icons";

const ScrollableApplicationDisclaimer: React.FunctionComponent = () => {
  return (
    <>
      <Card.Body>
        <Card.Title className="blue">
          CIIP Application Disclaimer
          <small>
            <a href="/resources/application-disclaimer" target="_blank">
              (<FontAwesomeIcon icon={faExternalLinkAlt} />
              expand)
            </a>
          </small>
        </Card.Title>
        <Card>
          <div tabIndex={0} id="disclaimer-text" className="show-scrollbar">
            <Card.Body>
              <LegalDisclaimerText />
            </Card.Body>
          </div>
        </Card>
      </Card.Body>
      <style jsx>{`
        #disclaimer-text {
          max-height: 19.2em;
          background: #eee;
          overflow-y: scroll;
        }
        .show-scrollbar::-webkit-scrollbar {
          -webkit-appearance: none;
          width: 8px;
        }
        .show-scrollbar::-webkit-scrollbar-thumb {
          border-radius: 4px;
          background-color: rgba(0, 0, 0, 0.5);
          box-shadow: 0 0 1px rgba(255, 255, 255, 0.5);
        }
        small {
          margin-left: 0.5em;
        }
      `}</style>
    </>
  );
};

export default ScrollableApplicationDisclaimer;
