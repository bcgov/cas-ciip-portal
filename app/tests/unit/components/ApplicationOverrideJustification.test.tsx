import React from 'react';
import {shallow} from 'enzyme';
import {ApplicationOverrideJustificationComponent} from 'components/Application/ApplicationOverrideJustification';

describe('OverrideJustification', () => {
  it('Should allow the user to save a justification if one is not currently active', async () => {
    const r = shallow(
      <ApplicationOverrideJustificationComponent
        hasErrors
        overrideActive={false}
        setOverrideActive={jest.fn()}
        applicationOverrideJustification={null}
        revisionId="abc"
        relay={null}
      />
    );
    expect(r).toMatchSnapshot();
    expect(r.find('AccordionCollapse').find('Button').text()).toBe('Save');
  });

  it('Should allow the user to edit a justification if one is currently active', async () => {
    const r = shallow(
      <ApplicationOverrideJustificationComponent
        overrideActive
        hasErrors
        setOverrideActive={jest.fn()}
        applicationOverrideJustification="bad stuff"
        revisionId="abc"
        relay={null}
      />
    );
    expect(r).toMatchSnapshot();
    expect(r.find('Alert').find('Button').at(0).text()).toBe(
      'Edit Justification'
    );
  });

  it('Should allow the user to delete a justification if one is currently active', async () => {
    const r = shallow(
      <ApplicationOverrideJustificationComponent
        overrideActive
        hasErrors
        setOverrideActive={jest.fn()}
        applicationOverrideJustification="bad stuff"
        revisionId="abc"
        relay={null}
      />
    );
    expect(r.find('Alert').find('Button').at(1).text()).toBe('Delete Override');
  });

  it('Should not render the override justification box if the application has no errors', async () => {
    const r = shallow(
      <ApplicationOverrideJustificationComponent
        overrideActive
        hasErrors={false}
        setOverrideActive={jest.fn()}
        applicationOverrideJustification="bad stuff"
        revisionId="abc"
        relay={null}
      />
    );
    expect(r.exists('Alert')).toBe(false);
    expect(r.exists('AccordionCollapse')).toBe(false);
  });
});
