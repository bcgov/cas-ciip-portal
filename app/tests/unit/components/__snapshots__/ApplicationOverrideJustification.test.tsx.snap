// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverrideJustification Should allow the user to edit a justification if one is currently active 1`] = `
<Fragment>
  <Alert
    closeLabel="Close alert"
    show={true}
    transition={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "appear": false,
          "in": false,
          "mountOnEnter": false,
          "timeout": 300,
          "unmountOnExit": false,
        },
        "displayName": "Fade",
        "render": [Function],
      }
    }
    variant="danger"
  >
    <strong>
      Error Validation Override Active
    </strong>
  </Alert>
  <Alert
    closeLabel="Close alert"
    show={true}
    transition={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "appear": false,
          "in": false,
          "mountOnEnter": false,
          "timeout": 300,
          "unmountOnExit": false,
        },
        "displayName": "Fade",
        "render": [Function],
      }
    }
    variant="secondary"
  >
    <p>
      <strong>
        You have chosen to override the errors present in your application. This may cause a delay in processing your application.
      </strong>
    </p>
    <p>
      <strong>
        Your Override Justification:
      </strong>
    </p>
    <p>
      bad stuff
    </p>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "marginRight": "5px",
        }
      }
      type="button"
      variant="secondary"
    >
      Edit Justification
    </Button>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      type="button"
      variant="danger"
    >
      Delete Override
    </Button>
  </Alert>
</Fragment>
`;

exports[`OverrideJustification Should allow the user to save a justification if one is not currently active 1`] = `
<Fragment>
  <Alert
    closeLabel="Close alert"
    show={true}
    transition={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "appear": false,
          "in": false,
          "mountOnEnter": false,
          "timeout": 300,
          "unmountOnExit": false,
        },
        "displayName": "Fade",
        "render": [Function],
      }
    }
    variant="danger"
  >
    <ForwardRef>
      <div
        className="jsx-905902600 override-accordion"
      >
        Your application contains errors shown below that must be fixed before submission. You may either correct these or alternatively, override and provide justification. Your justification will be reviewed by CAS and may cause a delay in processing your application.
        <ForwardRef
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "active": false,
                "disabled": false,
                "type": "button",
                "variant": "primary",
              },
              "displayName": "Button",
              "render": [Function],
            }
          }
          eventKey="0"
          onClick={[Function]}
          variant="secondary"
        >
          Override and Justify
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  320,
                  512,
                  Array [],
                  "f0d7",
                  "M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z",
                ],
                "iconName": "caret-down",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            style={
              Object {
                "marginLeft": "0.75em",
              }
            }
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </ForwardRef>
      </div>
      <AccordionCollapse
        eventKey="0"
        in={false}
      >
        <Form
          inline={false}
        >
          <h4
            className="jsx-905902600"
          >
            Override Form Validation
          </h4>
          <FormGroup
            controlId="overrideJustification"
          >
            <FormLabel
              column={false}
              srOnly={false}
            >
              Justification for incomplete form:
            </FormLabel>
            <FormControl
              as="textarea"
              onChange={[Function]}
              rows={4}
              value=""
            />
          </FormGroup>
          <Button
            active={false}
            disabled={false}
            onClick={[Function]}
            type="button"
            variant="success"
          >
            Save
          </Button>
          <ForwardRef
            as={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "active": false,
                  "disabled": false,
                  "type": "button",
                  "variant": "primary",
                },
                "displayName": "Button",
                "render": [Function],
              }
            }
            eventKey="0"
            onClick={[Function]}
            style={
              Object {
                "border": "1px solid currentColor",
                "marginLeft": "1em",
              }
            }
            variant="light"
          >
            Cancel
          </ForwardRef>
        </Form>
        <p
          className="jsx-905902600 justification-text"
        />
      </AccordionCollapse>
    </ForwardRef>
  </Alert>
  <JSXStyle
    id="905902600"
  >
    .override-accordion.jsx-905902600{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-wrap:wrap;-ms-flex-wrap:wrap;flex-wrap:wrap;-webkit-box-pack:end;-webkit-justify-content:flex-end;-ms-flex-pack:end;justify-content:flex-end;}.justification-text.jsx-905902600{color:red;}
  </JSXStyle>
</Fragment>
`;
